example/ionExample.js

```diff
@@ -0,0 +1,731 @@
+	AmbientLight,
+	WebGLRenderer,
+	PerspectiveCamera,
+	CameraHelper,
+	Raycaster,
+	Vector2,
+	Mesh,
+	CylinderBufferGeometry,
+	MeshBasicMaterial,
+	Group,
+	TorusBufferGeometry,
+	OrthographicCamera,
+	sRGBEncoding
+} from 'three';
+import { OrbitControls } from 'three/examples/jsm/controls/OrbitControls.js';
+import { BufferGeometryUtils } from 'three/examples/jsm/utils/BufferGeometryUtils.js';
+import { GLTFLoader } from 'three/examples/jsm/loaders/GLTFLoader.js';
+import { DRACOLoader } from 'three/examples/jsm/loaders/DRACOLoader.js';
+import * as dat from 'three/examples/jsm/libs/dat.gui.module.js';
+import Stats from 'three/examples/jsm/libs/stats.module.js';
+
+const ALL_HITS = 1;
+const FIRST_HIT_ONLY = 2;
+
+const hashUrl = window.location.hash.replace( /^#/, '' );
+let camera, controls, scene, renderer, tiles, cameraHelper;
+let thirdPersonCamera, thirdPersonRenderer, thirdPersonControls;
+let secondRenderer, secondCameraHelper, secondControls, secondCamera;
+let orthoCamera, orthoCameraHelper;
+let raycaster, mouse, rayIntersect, lastHoveredElement;
+let offsetParent;
+let statsContainer, stats;
+
+let params = {
+
+	'enableUpdate': true,
+	'raycast': NONE,
+	'enableCacheDisplay': false,
+	'enableRendererStats': false,
+	'orthographic': false,
+
+	'ionAssetId': '40866',
+	'ionAccessToken': 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJqdGkiOiIwY2Q2MzQ1OS1kNjI4LTRiZDEtOWVkZC1kMWI4YzAyODU3OGMiLCJpZCI6MjU5LCJpYXQiOjE2MDY4NzMyMTh9.8EwC6vilVHM2yizt8nG6VmbNu66QiCrk3O-1lEDPI9I',
+	'preFetchToken': false,
+	'errorTarget': 6,
+	'errorThreshold': 60,
+	'maxDepth': 15,
+	'loadSiblings': true,
+	'stopAtEmptyTiles': true,
+	'displayActiveTiles': false,
+	'resolutionScale': 1.0,
+
+	'up': hashUrl ? '+Z' : '+Y',
+	'displayBoxBounds': false,
+	'colorMode': 0,
+	'showThirdPerson': false,
+	'showSecondView': false,
+	'reload': reinstantiateTiles,
+
+};
+
+init();
+animate();
+
+function loadIonJson( assetId, accessToken ) {
+
+	const ionEndpointUrl = new URL(
+		`https://api.cesium.com/v1/assets/${assetId}/endpoint`
+	);
+	ionEndpointUrl.searchParams.append( "access_token", accessToken );
+
+	return fetch( ionEndpointUrl, { mode: 'cors' } )
+		.then( ( res ) => {
+
+			if ( res.ok ) {
+
+				return res.json();
+
+			} else {
+
+				return Promise.reject( `${res.status} : ${res.statusText}` );
+
+			}
+
+		} )
+		.then( ( json ) => {
+
+			return [ json.url, json.accessToken ];
+
+		} );
+
+}
+
+function setupTiles() {
+
+	tiles.fetchOptions.mode = 'cors';
+
+	tiles.manager.addHandler( /\.gltf$/, {
+
+		parse( ...args ) {
+
+			// Note the DRACO compression files need to be supplied via an explicit source.
+			// We use unpkg here but in practice should be provided by the application.
+			const dracoLoader = new DRACOLoader();
+			dracoLoader.setDecoderPath( 'https://unpkg.com/three@0.123.0/examples/js/libs/draco/gltf/' );
+			dracoLoader.setDecoderConfig( { type: "js" } ); // WASM overloads memory without some sort of bottleneck on loading
```

Comment:
I had to set draco loading to use javascript instead of WASM in the ionExample.js script. Because using WASM caused the sandboxed memory of my browser to run out. This could be solved by either calling the loading of gltf models via a bottlenecking function or doing it in workers.

Reply:
Interesting -- there are a couple things happening that might be leading to extra memory being created / kept around:

- We have to create a [duplicate buffer of just the GLTF bytes from the b3dm buffer](https://github.com/NASA-AMMOS/3DTilesRendererJS/blob/master/src/three/B3DMLoader.js#L17) because three.js' GLTFLoader does not account for a typed arrays byte offset or byte length. I suggested respecting the offset and length [here](https://github.com/mrdoob/three.js/pull/19616#issuecomment-642281533) but it hasn't been added, yet.

- Both the featureTable and batchTable [retain a reference to the full original B3DM buffer](https://github.com/NASA-AMMOS/3DTilesRendererJS/blob/master/src/base/B3DMLoaderBase.js#L70-L75) meaning that after draco decoding has unpacked the byte buffers we're still keeping around the originally loaded compressed buffer, as well as the new decompressed buffer. This can be fixed by copying the relevant feature and batch table data to new buffers which will cut down on the retained memory there.

I'm not sure if those two things are the source of this issue, though, particularly if using the js version of the decoder is working just fine. If you can track down an issue a bit more it might be worth opening an issue in the three.js repo.

Reply:
I think I found the source when I was glancing over the code -- the issue was that a new `DracoLoader` was being created for every tile being loaded which loaded a new instance of the WASM / JS code which also slowed down parsing a lot. Creating a single common DracoLoader prevents the draco code from being loaded multiple times and in turn greatly speeds up the loading and parsing of tiles. Here's a diff of the change I made to the use case example in the README: https://github.com/NASA-AMMOS/3DTilesRendererJS/commit/cb3328c9af176415ecb0620c05cb855573ccadca

Reply:
fixed it in my example.